cmd/clusterctl/client/repository/components.go

```diff
@@ -193,21 +193,40 @@ func NewComponents(provider config.Provider, version string, rawyaml []byte, con
 		return nil, errors.Wrap(err, "failed to perform variable substitution")
 	}
 
-	// transform the yaml in a list of objects, so following transformation can work on typed objects (instead of working on a string/slice of bytes)
+	return processComponentsYaml(provider, version, yaml, variables, configClient, targetNamespace, watchingNamespace)
+}
+
+// NewComponents returns a new objects embedding a component YAML file
+//
+// It is important to notice that clusterctl applies a set of processing steps to the “raw” component YAML read
+// from the provider repositories:
+// 1. Checks for all the variables in the component YAML file without replacing the values
+// 2. Ensure all the provider components are deployed in the target namespace (apply only to namespaced objects)
+// 3. Ensure all the ClusterRoleBinding which are referencing namespaced objects have the name prefixed with the namespace name
+// 4. Set the watching namespace for the provider controller
+// 5. Adds labels to all the components in order to allow easy identification of the provider objects
+func NewComponents(provider config.Provider, version string, rawyaml []byte, configClient config.Client, targetNamespace, watchingNamespace string) (*components, error) {
```

Comment:
Suggestion: I'm wondering if we could refactor and collapse the two methods `NewComponents` and `NewComponentsWithValues` into one method takes a struct like `ComponentsInput` that contains all the required arguments and one more called `replaceVariables bool`. If `replaceVariables` is set, then we call `replaceVariables(...)`.

Reply:
I tried at first adding a bool like you are suggesting, but all kinds of tests broke because these functions are called from a bunch of places at different levels, so some places call replace variables directly and other places call NewComponents. It would have been a real big refactor

Reply:
Hmm...yes it would be a sizeable change across the layers of clusterctl from `getComponentsByName` -> `ComponentsClient.Get` -> `NewComponents`. But I'm leaning more towards that because currently the clusterctl API is subject to change regardless and I'd prefer not having extra methods in the interfaces that we may have to support in the future.

Reply:
these are different methods that are used for different use cases, adding the bool won't change that, it would just make header to read

Reply:
As per comment https://github.com/kubernetes-sigs/cluster-api/issues/2876#issuecomment-613909426, I don't see such a clear distinction between the two use cases because variable substitution could provide value to users also in `clusterctl config cluster` with the `-o yaml` option.

Said that I like the idea of avoiding duplicating methods, even if I fully understand the pain of fixing the chain of calls (I experimented also some refactoring to get rid of this, but this is invasive).